site/content/protocols/messagepickup/4.0/readme.md

```diff
@@ -0,0 +1,388 @@
+
+### Status
+Status details about waiting messages.
+
+Message Type URI: `https://didcomm.org/message-pickup/4.0/status`
+
+DIDComm v1 example:
+```json
+{
+    "@id": "123456780",
+    "@type": "https://didcomm.org/message-pickup/4.0/status",
+    "~thread": {
+        "thid": "<message id of status-request message>"
+    },
+    "recipient_did": "<did:key for messages>",
+    "message_count": 7,
+    "longest_waited_seconds": 3600,
+    "newest_received_time": 1658085169,
+    "oldest_received_time": 1658084293,
+    "total_bytes": 8096,
+    "live_delivery": false
+}
+```
+
+DIDComm v2 example:
+```json
+{
+    "id": "123456780",
+    "thid": "<message id of status-request message>",
+    "type": "https://didcomm.org/message-pickup/4.0/status",
+    "body": {
+        "recipient_did": "<did for messages>",
+        "message_count": 7,
+        "longest_waited_seconds": 3600,
+        "newest_received_time": 1658085169,
+        "oldest_received_time": 1658084293,
+        "total_bytes": 8096,
+        "live_delivery": false
+    }
+}
+```
+
+`message_count` is the only **REQUIRED** attribute. The others **MAY** be present if offered by the `mediator`.
+
+`longest_waited_seconds` is in seconds, and is the longest delay of any message in the queue.
+
+`newest_received_time` and `oldest_received_time` are expressed in UTC Epoch Seconds (seconds since 1970-01-01T00:00:00Z) as an integer.
+
+`total_bytes` represents the total size of all messages.
+
+If a `recipient_did` was specified in the `status-request` message, the matching value **MUST** be specified in the `recipient_did` attribute of the status message.
+
+`live_delivery` state is also indicated in the status message.
+
+**Note**: due to the potential for confusing what the actual state of the message queue is, a `status` message **MUST NOT** be put on the pending message queue and **MUST** only be sent when the `recipient` is actively connected (HTTP request awaiting response, WebSocket, etc.).
+
+
+### Delivery Request
+A request from the `recipient` to the `mediator` to have pending messages delivered.
+
+Message Type URI: `https://didcomm.org/message-pickup/4.0/delivery-request`
+
+DIDComm v1 example:
+```json
+{
+    "@id": "123456780",
+    "@type": "https://didcomm.org/message-pickup/4.0/delivery-request",
+    "limit": 10,
+    "recipient_did": "<did:key for messages>",
+    "~transport": {
+        "return_route": "all"
+    }
+}
+```
+
+DIDComm v2 example:
+```json
+{
+    "id": "123456780",
+    "type": "https://didcomm.org/message-pickup/4.0/delivery-request",
+    "body": {
+        "limit": 10,
+        "recipient_did": "<did for messages>"
+    },
+    "return_route": "all"
+}
+```
+
+`limit` is a **REQUIRED** attribute, and after receipt of this message, the `mediator` **SHOULD** deliver up to the limit indicated.
```

Comment:
Since there is some message exchange overhead for every loop (delivery-request -> delivery -> messages-received -> status), I think it is important to keep loop number as minimal as possible. Is actually a limit by message number the best approach, or should we consider also the transport capabilities of the recipient (i.e. maximum single message size)?

A recipient would likely want to receive all messages at once, as long as they fit into a single message. So I'm wondering that we can combine this specified `limit` with the `max_receive_bytes` [constraint](https://identity.foundation/didcomm-messaging/spec/#max_receive_bytes), in such a way that the mediator should consider both when constructing each delivery message.

I think this could work nicely unless a certain queued message exceeds recipient's capabilities: in such case, the message would be stuck in the queue indefinitely. Shall the mediator reject any inbound message for a recipient if it exceeds its reported `max_receive_bytes` value?

Reply:
> A recipient would likely want to receive all messages at once, as long as they fit into a single message. So I'm wondering that we can combine this specified limit with the max_receive_bytes [constraint](https://identity.foundation/didcomm-messaging/spec/#max_receive_bytes), in such a way that the mediator should consider both when constructing each delivery message.

My only concern here is regarding additional processing power/resources required on the `mediator` side to determine the number of messages attachable. Is the `max_receive_bytes` considered by the raw message size, or after packing for delivery (inside a delivery wrapper, and then inside a forward wrapper)?

> I think this could work nicely unless a certain queued message exceeds recipient's capabilities: in such case, the message would be stuck in the queue indefinitely. Shall the mediator reject any inbound message for a recipient if it exceeds its reported max_receive_bytes value?

Perhaps that should actually be covered in the Mediation Coordination protocol? Since it sounds like it should be set up once and forgotten. But I do agree having some mechanism for indicating what is too large of a message is probably ideal.

Reply:
I think you may be right @genaris. @TheTechmage do you know if this would be a lot of work to add to existing mediator implementations? I lean towards having it be an optional field (vs required) and then having the number of messages be determined for whichever is lower (limit or max_received_bytes).